KAFKA-12879: Revert changes from KAFKA-12339 and instead add retry capability to KafkaBasedLog

philipnee opened the pull request:
**Summary**
Fixes the compatibility issue regarding KAFKA-12879 by reverting the changes to the admin client from KAFKA-12339 (#10152), and instead perform the retries within the KafkaBasedLog via a new `TopicAdmin.retryEndOffsets(..)`.  This method will delegate to the existing `TopicAdmin.endOffsets(...)` and will retry on RetriableException until the _retries_ (max retries) is hit, or until a specified timeout.  

This change should be backward compatible to the KAFKA-12339.

Notable changes are:
**Reverted KAFKA-12339**

**TopicAdmin**
Added new `retryEndOffsets(...)` method that utilizes the new `RetryUtil` to list end offsets for a set of topic partitions and perform retries upon any RetriableException (including UnknownTopicOrPartitionException, TimeoutException).

**KafkaBasdLog**
During startup calls `TopicAdmin.retryEndOffsets(...)` to read end offsets for the log topic and retries if the topic is not available (or any RetriableException).

**RetryUtil**
A general utility that takes a Callable function, the total duration to retry (may be 0), and backoff time in ms, and that calls the function at least once, and performs retries using the parameters when the function throws RetriableException (e.g. UnknownTopicOrPartitionException, TimeoutException)

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

philipnee committed:
initial commit - wip

philipnee committed:
wip - tests

philipnee committed:
Clean up commits

philipnee committed:
RetryUtil

philipnee committed:
clean up

philipnee committed:
clean up tests

philipnee committed:
Revert KAFKA-12339 (#10152)

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:
```diff
@@ -703,6 +706,20 @@ public Config describeTopicConfig(String topic) {
+        try {
+            return RetryUtil.retry(
+                    () -> endOffsets(partitions),
+                    maxRetries,
```
The values of `maxRetries` and `retryBackoffMs` can each be `0`, and these are passed directly into the `RetryUtil.retry(...)` method. See my comment there.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/RetryUtil.java:
```diff
@@ -0,0 +1,52 @@
+        int retries = 0;
+        while (retries++ < maxRetries) {
+            try {
+                return callable.call();
```
Even though this method will not be part of our public API, we should still add JavaDoc so that we can document what the allowed values for `maxRetries` and `retryBackoffMs` can be, and what exceptions might be thrown by this method.

Second, is `maxRetries` the total number of _attempts_, or the total number of _retries_ that are allowed. With the way this method is called by `KafkaBasedLog`, it is possible that both or either of these values are `0` -- which makes sense for "retries". But if `retries` is `0`, then this method will retry for a long period of times. Let's fix this logic to handle either or both `maxRetries` and/or `retryBackoffMs` being `0`, and add unit tests to covert this situation.

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/RetryUtilTest.java:
```diff
@@ -0,0 +1,107 @@
+    @Test
+    public void TestSuccess() throws Exception {
+        Mockito.when(mockCallable.call()).thenReturn("success");
+        //EasyMock.expect(mockCallable.call()).andReturn("success").times(1);
```
Let's remove this line

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/RetryUtilTest.java:
```diff
@@ -0,0 +1,107 @@
+    public void TestSuccess() throws Exception {
+        Mockito.when(mockCallable.call()).thenReturn("success");
+        //EasyMock.expect(mockCallable.call()).andReturn("success").times(1);
+        try {
+            assertEquals("success", RetryUtil.retry(mockCallable, 10, 100));
+        } catch (Exception e) {
+            fail("Not expecting an exception: ", e.getCause());
+        }
```
Is this any better than just:
```suggestion
        assertEquals("success", RetryUtil.retry(mockCallable, 10, 100));
```

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/RetryUtilTest.java:
```diff
@@ -0,0 +1,107 @@
+    @Test
+    public void TestFailingRetries() throws Exception {
+        Mockito.when(mockCallable.call()).thenThrow(new TimeoutException());
+        try {
+            RetryUtil.retry(mockCallable, 10, 100);
+            fail("Expect exception being thrown here");
+        } catch (ConnectException e) {
+            // expecting a connect exception
+        } catch (Exception e) {
+            fail("Only expecting ConnectException");
+        }
```
It's much better to use `Assertion.assertThrows(...)`:
```suggestion
        ConnectException e = assertThrows(ConnectException.class, RetryUtil.retry(mockCallable, 10, 100));
        assertTrue(e.getMesssage().startsWith("Fail to retry the operation after 10 attempts."));
```

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/RetryUtilTest.java:
```diff
@@ -0,0 +1,107 @@
+                .thenThrow(new TimeoutException("timeout"))
+                .thenThrow(new TimeoutException("timeout"))
+                .thenThrow(new NullPointerException("Non retriable"));
+        try {
+            for (int i = 0; i < 10; i++) {
+                RetryUtil.retry(mockCallable, 10, 100);
+            }
+            fail("Not expecting an exception: ");
+        } catch (ConnectException e) {
+            fail("Should fail with NPE");
+        } catch (NullPointerException e) {
+            // good
+        }
```
What is the purpose of the for loop? Why not just:
```suggestion
        NullPointerException e = assertThrows(NullPointerException.class, RetryUtil.retry(mockCallable, 10, 100));
        assertTrue(e.getMesssage().startsWith("Non retriable"));
```

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/RetryUtilTest.java:
```diff
@@ -0,0 +1,107 @@
+                .thenThrow(new TimeoutException())
+                .thenThrow(new TimeoutException())
+                .thenReturn("success");
+        try {
+            assertEquals("success", RetryUtil.retry(mockCallable, 10, 100));
+        } catch (Exception e) {
+            fail("Not expecting an exception: ", e.getCause());
+        }
```
The try-catch is not really needed, since any exception thrown will fail the test:
```suggestion
        assertEquals("success", RetryUtil.retry(mockCallable, 10, 100));
```

rhauch reviewed (commented):
Thanks, @philipnee. I have a few comments below, but otherwise this PR is pretty close.

philipnee committed:
Address PR comments

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/KafkaBasedLog.java:
```diff
@@ -319,9 +319,11 @@ private void poll(long timeoutMs) {
+    private void readToLogEnd(boolean shouldRetry) {
         Set<TopicPartition> assignment = consumer.assignment();
-        Map<TopicPartition, Long> endOffsets = readEndOffsets(assignment);
+        // readEndOffsets makes listOffsets call to adminClient, if shouldRetry is set to True, the adminClinet
+        // will retry on RetriableExceptions
```
Might be worth moving this to a JavaDoc for the method.

philipnee replied:
Yes! Good call.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/RetryUtil.java:
```diff
@@ -0,0 +1,72 @@
+    public static <T> T retry(Callable<T> callable, long maxRetries, long retryBackoffMs) throws Exception {
+        Throwable lastError = null;
+        int attempt = 0;
+        long maxAttempts = maxRetries + 1;
```
Nit:
```suggestion
        final long maxAttempts = maxRetries + 1;
```

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/RetryUtil.java:
```diff
@@ -0,0 +1,72 @@
+        Throwable lastError = null;
+        int attempt = 0;
+        long maxAttempts = maxRetries + 1;
+        while (attempt++ < maxAttempts) {
```
Should this be:
```suggestion
        while (++attempt <= maxAttempts) {
```
so we could use a 1-based attempt in the log message on line 58-59?

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/RetryUtil.java:
```diff
@@ -0,0 +1,72 @@
+            try {
+                return callable.call();
+            } catch (RetriableException | org.apache.kafka.connect.errors.RetriableException e) {
+                log.warn("RetriableException caught, retrying automatically up to {} more times. " +
+                        "Reason: {}", maxRetries - attempt, e.getMessage());
```
WDYT?
```suggestion
                log.warn("RetriableException caught on attempt {}, retrying automatically up to {} more times. " +
                        "Reason: {}", attempt, maxAttempts - attempt, e.getMessage());
```

rhauch reviewed (commented):
Thanks, @philipnee. I have a few more minor comments below.

philipnee committed:
PR comment

- javaDoc
- better logging

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/KafkaBasedLog.java:
```diff
@@ -319,9 +319,18 @@ private void poll(long timeoutMs) {
+     * As the <code>listOffsets</code> method might throw a {@link RetriableException}, the <code>shouldRetry</code>
+     * flag enables retry, upon catching such exception, if it is set to <code>True</code>.
+     *
+     * @param shouldRetry Boolean flag to enable retry for the admin client <code>listOffsets</code> call.
```
Maybe add:
```suggestion
     * @param shouldRetry Boolean flag to enable retry for the admin client <code>listOffsets</code> call.
     * @see TopicAdmin#retryEndOffsets
```

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/KafkaBasedLog.java:
```diff
@@ -319,9 +319,18 @@ private void poll(long timeoutMs) {
+    private void readToLogEnd(boolean shouldRetry) {
         Set<TopicPartition> assignment = consumer.assignment();
-        Map<TopicPartition, Long> endOffsets = readEndOffsets(assignment);
+        // it will subsequently invoke the listOffsets call here
```
I don't think this comment adds much value, especially because "subsequently" is ambiguous. IMO the method call stands on its own.
```suggestion
```

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/RetryUtil.java:
```diff
@@ -0,0 +1,72 @@
+    private static final Logger log = LoggerFactory.getLogger(RetryUtil.class);
+
+    /**
+     * The method executes the callable, and performs retries if
+     * {@link org.apache.kafka.connect.errors.RetriableException} is being thrown.  If other types of exceptions is
+     * caught, then the same exception will be rethrown.  If all retries are exhausted, then the last
+     * exception is wrapped into a {@link org.apache.kafka.connect.errors.ConnectException} and rethrown.
+     *
+     * The callable task will be executed at least once.  If <code>maxRetries</code> is set to 0, the task will be
+     * executed exactly once.  If <code>maxRetries</code> is set to <code>n</code>, the callable will be executed at
+     * most <code>n + 1</code> times.
+     *
+     * If <code>retryBackoffMs</code> is set to 0, no wait will happen in between the retries.
```
A few corrections to the JavaDoc:
```suggestion
     * The method executes the callable at least once, optionally retrying the callable if
     * {@link org.apache.kafka.connect.errors.RetriableException} is thrown. All other types of exceptions
     * are not caught and are rethrown. If all retries are exhausted, the last
     * exception is wrapped with a {@link org.apache.kafka.connect.errors.ConnectException} and thrown.
     *
     * <p>If <code>maxRetries</code> is set to 0, the task will be executed exactly once.
     * If <code>maxRetries</code> is set to <code>n</code>, the callable will be executed at
     * most <code>n + 1</code> times.
     *
     * <p>If <code>retryBackoffMs</code> is set to 0, no wait will happen in between the retries.
```

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/RetryUtil.java:
```diff
@@ -0,0 +1,72 @@
+     *
+     * If <code>retryBackoffMs</code> is set to 0, no wait will happen in between the retries.
+     *
+     * @param callable The task to execute.
+     * @param maxRetries Maximum number of retries.
+     * @param retryBackoffMs Delay time to retry the callable task upon receiving a
+     * {@link org.apache.kafka.connect.errors.RetriableException}.
```
```suggestion
     * @param callable       the function to execute.
     * @param maxRetries     maximum number of retries; must be 0 or more
     * @param retryBackoffMs the number of milliseconds to delay upon receiving a
     *.                      {@link org.apache.kafka.connect.errors.RetriableException}
     *                       before retrying again; must be 0 or more
```

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/RetryUtil.java:
```diff
@@ -0,0 +1,72 @@
+            Utils.sleep(retryBackoffMs);
+        }
+
+        throw new ConnectException("Fail to retry the task after " + maxRetries + " attempts.  Reason: " + lastError, lastError);
```
Suggestion to improve the message:
```suggestion
        throw new ConnectException("Fail to retry the task after " + maxAttempts + " attempts.  Reason: " + lastError.getMessage(), lastError);
```

And, if `maxRetries == 0` should we just call the function without any special handling, since we're not retrying? For example, should we add something like this very early in the method? Doing that would make it easier to phrase this message, since the `ConnectException` will only be used when at least 1 retry may be attempted.
```
if (maxRetries <= 0) {
    // no special error handling
    return callable.call();
}

philipnee replied:
Agreed.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/RetryUtil.java:
```diff
@@ -0,0 +1,72 @@
+                log.warn("Non-retriable exception caught. Re-throwing. Reason: {}, {}", e.getClass(), e.getMessage());
+                throw e;
+            }
+            Utils.sleep(retryBackoffMs);
```
```suggestion
            if (attempt < maxAttempts) {
                Utils.sleep(retryBackoffMs);
            }
```

philipnee replied:
Agreed, and I actually had the exact same though - though, i thought it would be fine to wait for an additional x-amount of time.

rhauch replied:
Yeah, I think it's worth the bit of logic to fail more quickly.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/RetryUtil.java:
```diff
@@ -0,0 +1,72 @@
+                lastError = e;
+            } catch (WakeupException e) {
+                lastError = e;
+            } catch (Exception e) {
+                log.warn("Non-retriable exception caught. Re-throwing. Reason: {}, {}", e.getClass(), e.getMessage());
+                throw e;
```
Do we need this catch block and log message? We're re-throwing the exception, and the stack trace will include the fact that the function threw the exception, and the log message here doesn't seem to add much detail since it will be caught and dealt with by the caller.
```suggestion
```
If we remove this, we should adjust the JavaDoc to not mention "rethrow" since we would not catch the problem and are not rethrowing the exception.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/RetryUtil.java:
```diff
@@ -0,0 +1,72 @@
+            try {
+                return callable.call();
+            } catch (RetriableException | org.apache.kafka.connect.errors.RetriableException e) {
+                log.warn("RetriableException caught on attempt {}, retrying automatically up to {} more times. " +
+                        "Reason: {}", attempt, maxRetries - attempt, e.getMessage());
```
Are we intentionally not logging the exception as the extra parameter? If the exception wraps a more useful exception, we won't see any information about the wrapped exception unless we can see the stack trace in the warning log message.

philipnee replied:
We probably should log the exception, in fact.  making the change.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:
```diff
@@ -703,6 +706,20 @@ public Config describeTopicConfig(String topic) {
     }
 
+    protected Map<TopicPartition, Long> retryEndOffsets(Set<TopicPartition> partitions) {
+        int maxRetries = Integer.parseInt(adminConfig.getOrDefault(AdminClientConfig.RETRIES_CONFIG, DEFAULT_ADMIN_CLIENT_RETRIES).toString());
```
Maybe add a comment here:
```suggestion
        // These will be 0 or higher, enforced by the admin client when it is instantiated and configured
        int maxRetries = Integer.parseInt(adminConfig.getOrDefault(AdminClientConfig.RETRIES_CONFIG, DEFAULT_ADMIN_CLIENT_RETRIES).toString());
```

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/RetryUtilTest.java:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    @Test
+    public void TestSuccess() throws Exception {
```
Nit: all of these test methods should start with a lowercase character.

rhauch reviewed (commented):
Thanks again, @philipnee. Getting to very nitty details now, but I think if all of these are addressed then we should be good.

philipnee committed:
Update PR

- Updated documentation
- Updated log message in the retry util, now getting ```[2022-02-25 13:18:34,313] WARN RetriableException caught on attempt 1, retrying automatically up to 9 more times. Reason: timeout exception (org.apache.kafka.connect.util.RetryUtil:62)
org.apache.kafka.common.errors.TimeoutException: timeout exception```
- Improved RetryUtil logic: no retry if maxRetries = 0, no extra sleep for the last retry attempt
- lower case for test methods

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/RetryUtil.java:
```diff
@@ -0,0 +1,76 @@
+     *
+     * <p>If <code>retryBackoffMs</code> is set to 0, no wait will happen in between the retries.
+     *
+     * @param callable the function to execute.
+     * @param maxRetries maximum number of retries; must be 0 or more
+     * @param retryBackoffMs the number of milliseconds to delay upon receiving a
+     * {@link org.apache.kafka.connect.errors.RetriableException} before retrying again; must be 0 or more
+     *
```
Nit on the spacing so the description of parameters is column-aligned.
```suggestion
     * @param callable       the function to execute.
     * @param maxRetries     maximum number of retries; must be 0 or more
     * @param retryBackoffMs the number of milliseconds to delay upon receiving a
     *                       {@link org.apache.kafka.connect.errors.RetriableException} before retrying again; 
     *                       must be 0 or more
```

rhauch replied:
@philipnee can you please correct this spacing to reflect the project standards? Thanks!

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/RetryUtilTest.java:
```diff
@@ -0,0 +1,127 @@
+@RunWith(PowerMockRunner.class)
+public class RetryUtilTest {
+
+    @Mock
```
Do we need this line? I think not since you're instantiating `mockCallable` in `setUp()`. The annotation is really only needed when the Mockito JUnit runner is used to initialize the mock fields.

Also, IIRC we get rid of the `@RunWith(PowerMockRunner.class)` line as well, since this code is not really using anything from PowerMock. It'd be great if we could avoid using PowerMock in new code.

philipnee replied:
oh good call - the mock exist because I was doing something else.  Deleting this line.

rhauch reviewed (commented):
Okay, just a few more nits related to your recent changes.

philipnee committed:
PR Revision

- minors

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/KafkaBasedLog.java:
```diff
@@ -319,9 +319,18 @@ private void poll(long timeoutMs) {
+     * @param shouldRetry Boolean flag to enable retry for the admin client {@code listOffsets()} call.
+     * @see TopicAdmin#retryEndOffsets
+     */
+
```
Nit:
```suggestion
```

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/RetryUtil.java:
```diff
@@ -0,0 +1,76 @@
+     * then the last exception is wrapped with a {@link org.apache.kafka.connect.errors.ConnectException} and thrown.
+     *
+     * <p>If {@code maxRetries} is set to 0, the task will be
+     * executed exactly once.  If {@code maxRetries} is set to ,{@code n} the callable will be executed at
```
Nit:
```suggestion
     * executed exactly once.  If {@code maxRetries} is set to {@code n}, the callable will be executed at
```

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/KafkaBasedLog.java:
```diff
@@ -319,9 +319,18 @@ private void poll(long timeoutMs) {
+    /**
+     * This method finds the end offsets using the {@code listOffsets()} method of the admin client.
+     * As the {@code listOffsets()} method might throw a {@link RetriableException}, the {@code shouldRetry}
+     * flag enables retry, upon catching such exception, if it is set to {@code True}.
```
Suggestions:
```suggestion
     * This method finds the end offsets of the Kafka log's topic partitions, optionally retrying
     * if the {@code listOffsets()} method of the admin client throws a {@link RetriableException}.
```

philipnee committed:
Update connect/runtime/src/main/java/org/apache/kafka/connect/util/RetryUtil.java

accepting comments on spacing

Co-authored-by: Randall Hauch <rhauch@gmail.com>

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:
```diff
@@ -124,6 +124,9 @@ public String toString() {
     public static final int NO_PARTITIONS = -1;
     public static final short NO_REPLICATION_FACTOR = -1;
 
+    private static final String DEFAULT_ADMIN_CLIENT_RETRIES = "10";
+    private static final String DEFAULT_ADMIN_CLIENT_BACKOFF_MS = "100";
+
```
Okay, so I was verifying what these defaults really were on the `AdminClientConfig`, and I noticed that [AdminClientConfig.RETRIES_CONFIG](https://github.com/apache/kafka/blob/2ccc834faa3fffcd5d15d2463aeef3ee6f5cea13/clients/src/main/java/org/apache/kafka/clients/admin/AdminClientConfig.java#L173-L178) defaults `Int.MAX_VALUE`. I did a bit of research, and found [KIP-533](https://cwiki.apache.org/confluence/display/KAFKA/KIP-533%3A+Add+default+api+timeout+to+AdminClient) changed the defaults, and I don't know that we can/should count on `AdminClientConfig.RETRIES`.

I consulted with @kkonstantine, and after a bit of discussion we agree that it would be better to not make these configurable and to instead define the maximum duration to try to read the offsets as a constant (e.g., 15 minutes), plus a constant for retry backoff millis (e.g., 10 seconds). The max duration is easier to reason about.

The 15 minutes is a balance between retrying long enough to work with most incidents of metadata propagation taking too long, and short enough to fail if the worker cannot start up.

Interestingly, I don't think the duration needs to be defined in `TopicAdmin`, and is better defined in `KafkaBasedLog.start()`. That way, the `KafkaBasedLog` passes the duration into the `TopicAdmin.endOffset(Set<TopicPartition> partitions, Duration timeout)` method, which is the method that calls the `RetryUtil.retry(Callable<T> callable, Duration retryTimeout, long retryBackoffMs)`.

This isn't too much different than what you have, but I do think it's cleaner. I don't think we have seen too many retriable exception during `KafkaBasedLog.start()`, which means it's kind of an edge case.

WDYT?



Second, it might be good to add these methods:
```
public Map<TopicPartition, Long> retryEndOffsets(Set<TopicPartition> partitions, Duration timeout) { ... }
```
that retries for the given timeout while using `DEFAULT_ADMIN_CLIENT_BACKOFF_MS` for backoffs. Then, if you change `KafkaBasedLog.readEndOffsets(...)` and `e the `KafkaBasedLog.readToLogEnd(...)` to take a `Duration timeout` rather than a boolean, then `KafkaBasedLog.start() could just call `readToLogEnd(Duration.ofMinutes(15))` while `KafkaBasedLog.WorkThread.run()` could call `readToLogEnd(null)`.

You'd need to refactor your `retry(...)` method a bit to take the `Duration`. You could also

rhauch reviewed (commented):
Thanks, @philipnee. A few more nits below, but one bigger suggestion around how we know how long to retry. I know I originally suggested the approach in this PR, but I do think my suggested changes are going to be much better.

philipnee committed:
Apply suggestions from code review

Co-authored-by: Randall Hauch <rhauch@gmail.com>

philipnee committed:
Retry until timeout instead of exhausting number of retries

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/KafkaBasedLog.java:
```diff
@@ -76,6 +76,10 @@
+    // 15min of admin retry duration to ensure successful metadata propagation.  10 seconds of backoff
+    // in between retries
+    private static final Duration ADMIN_CLIENT_RETRY_DURATION = Duration.ofMinutes(15);
+    private static final long ADMIN_CLIENT_RETRY_BACKOFF_MS = 10 * 1000;
```
Nit: our convention is to use either `TimeUnit` or `Duration` to compute the millis to make it more readable and to reduce risk of conversion errors. Using `Duration` is a bit more readable, but the lines above already use `TimeUnit`:
```suggestion
    private static final long ADMIN_CLIENT_RETRY_BACKOFF_MS = TimeUnit.SECONDS.toMillis(10);
```

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/RetryUtil.java:
```diff
@@ -0,0 +1,87 @@
+     * <p>If {@code retryBackoffMs} is set to 0, no wait will happen in between the retries.
+     *
+     * @param callable          the function to execute.
+     * @param timeoutDuration   timeout duration
```
Nit: since we require a non-null `Duration`, we should state that here:
```suggestion
     * @param timeoutDuration   timeout duration; may not be null
```

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/RetryUtil.java:
```diff
@@ -0,0 +1,87 @@
+    public static <T> T retryUntilTimeout(Callable<T> callable, Duration timeoutDuration, long retryBackoffMs) throws Exception {
+        long timeoutMs = timeoutDuration.toMillis();
+        if (retryBackoffMs >= timeoutMs) {
+            log.warn("retryBackoffMs, {}, needs to be less than timeoutMs, {}.  Callable will only execute once.", retryBackoffMs, timeoutMs);
```
All of the log messages in this method could use a description of the callable. WDYT about adding a `Supplier<String> description` parameter, and using that in the log messages. For example, the `KafkaBasedLog` could supply 
```
() -> "list offsets for topic '" + topicName + "'"
```
then this log message might be:

```suggestion
            log.warn("Call to {} will only execute once, since retryBackoffMs={} is larger than total timeoutMs={}"),
                description.get(), retryBackoffMs, timeoutMs);
```

philipnee replied:
I like this idea.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/RetryUtil.java:
```diff
@@ -0,0 +1,87 @@
+            try {
+                return callable.call();
+            } catch (RetriableException | org.apache.kafka.connect.errors.RetriableException e) {
+                log.warn("RetriableException caught on attempt {}, retrying automatically. " +
+                        "Reason: {}", attempt, e.getMessage(), e);
```
Likewise, this log message could be changed to:
```suggestion
                log.warn("Attempt {} to {} resulted in RetriableException; retrying automatically. " +
                        "Reason: {}", attempt, description.get(), e.getMessage(), e);
```

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/RetryUtil.java:
```diff
@@ -0,0 +1,87 @@
+            }
+        }
+
+        throw new ConnectException("Fail to retry the task after " + attempt + " attempts.  Reason: " + lastError.getMessage(), lastError);
```
And this exception message could also use the description:
```suggestion
        throw new ConnectException("Fail to " + description.get() + " after " + attempt + " attempts.  Reason: " + lastError.getMessage(), lastError);
```

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/RetryUtil.java:
```diff
@@ -0,0 +1,87 @@
+
+            // if current time is less than the ending time, no more retry is necessary
+            // won't sleep if retryBackoffMs equals to 0
+            if (retryBackoffMs > 0 && System.currentTimeMillis() < end) {
+                Utils.sleep(retryBackoffMs);
+            }
```
WDYT about this:
```suggestion
            long millisRemaining = Math.max(0, end - System.currentTimeMillis());
            if (millisRemaining > 0) {
                Utils.sleep(millisRemaining)
            }
```

philipnee replied:
much better

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:
```diff
@@ -124,6 +124,9 @@ public String toString() {
     public static final int NO_PARTITIONS = -1;
     public static final short NO_REPLICATION_FACTOR = -1;
 
+    private static final String DEFAULT_ADMIN_CLIENT_RETRIES = "10";
+    private static final String DEFAULT_ADMIN_CLIENT_BACKOFF_MS = "100";
+
```
These are unused now, right?

philipnee replied:
ah 🤦

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/TopicAdminTest.java:
```diff
@@ -465,6 +467,55 @@ public void verifyingGettingTopicCleanupPolicies() {
+    }
+
+    @Test
+    public void endOffsetsShouldRetryWhenTopicNotFound() {
```
Nit:
```suggestion
    public void retryEndOffsetsShouldRetryWhenTopicNotFound() {
```

rhauch reviewed (commented):
Very nice, @philipnee! A few nits below, but otherwise looks good.

philipnee committed:
Apply suggestions from code review

Co-authored-by: Randall Hauch <rhauch@gmail.com>

philipnee committed:
Added message supplier to supply custom message

philipnee committed:
less retries to speed up the process

philipnee committed:
timeout test if unable to complete within a reasonable timeframe

philipnee committed:
Clean up the documentation

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/RetryUtil.java:
```diff
@@ -0,0 +1,105 @@
+            try {
+                return callable.call();
+            } catch (RetriableException | org.apache.kafka.connect.errors.RetriableException e) {
+                log.warn("Attempt {} to execute {} resulted in RetriableException; retrying automatically. " +
```
Nit: including "execute" here is completely unnecessary.
```suggestion
                log.warn("Attempt {} to {} resulted in RetriableException; retrying automatically. " +
```

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/RetryUtil.java:
```diff
@@ -0,0 +1,105 @@
+            }
+        }
+
+        throw new ConnectException("Fail to execute " + descriptionStr + " after " + attempt + " attempts.  Reason: " + lastError.getMessage(), lastError);
```
Nit: including "execute" here is completely unnecessary.
```suggestion
        throw new ConnectException("Fail to " + descriptionStr + " after " + attempt + " attempts.  Reason: " + lastError.getMessage(), lastError);
```

philipnee replied:
A comment on the logging message, I think they can be grammatically strange if the provided description isn't a verb. 
 E.g., "Fail to logging error". Probably not a big deal.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/RetryUtil.java:
```diff
@@ -0,0 +1,105 @@
+            // won't sleep if retryBackoffMs equals to 0
+            long millisRemaining = Math.max(0, end - System.currentTimeMillis());
+            if (millisRemaining > 0) {
+                Utils.sleep(retryBackoffMs);
```
What happens if `millisRemaining` is, say, 2 and `retryBackoffMs` is 1000? If `millisRemaining` is positive, then shouldn't we sleep for the smaller of `millisRemaining` or `retryBackoffMs`? IOW:
```suggestion
                Utils.sleep(Math.min(retryBackoffMs, millisRemaining));
```

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:
```diff
@@ -703,6 +703,19 @@ public Config describeTopicConfig(String topic) {
 
+    protected Map<TopicPartition, Long> retryEndOffsets(Set<TopicPartition> partitions, Duration timeoutDuration, long retryBackoffMs) {
+
+        try {
+            return RetryUtil.retryUntilTimeout(
+                    () -> endOffsets(partitions),
+                    () -> "list offsets",
+                    timeoutDuration,
+                    retryBackoffMs);
+        } catch (Exception e) {
+            throw new ConnectException("Failed to read offsets for topic partitions.", e);
+        }
```
We should use consisting phrasing, and actually "list offsets for topic partitions" most closely aligns with the admin client call that `endOffsets(...)` is using. So how about:
```suggestion
        try {
            return RetryUtil.retryUntilTimeout(
                    () -> endOffsets(partitions),
                    () -> "list offsets for topic partitions",
                    timeoutDuration,
                    retryBackoffMs);
        } catch (Exception e) {
            throw new ConnectException("Failed to list offsets for topic partitions.", e);
        }
```
Also, using `list offsets for topic partitions` as the retry function description works quite well with the log and exception message within `retryUntilTimeout(...)` method.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:
```diff
@@ -703,6 +703,19 @@ public Config describeTopicConfig(String topic) {
         return result;
     }
 
+    protected Map<TopicPartition, Long> retryEndOffsets(Set<TopicPartition> partitions, Duration timeoutDuration, long retryBackoffMs) {
```
Why is this protected when `endOffsets(...)` is public? This class is the Connect runtime module, and the class is not part of the public API for Connect.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/RetryUtil.java:
```diff
@@ -0,0 +1,105 @@
+        long timeoutMs = timeoutDuration.toMillis();
+
+        if (retryBackoffMs >= timeoutMs) {
+            log.warn("Call to {} will only execute once, since retryBackoffMs={} is larger than total timeoutMs={}",
+                    descriptionStr, retryBackoffMs, timeoutMs);
```
Nit:
```suggestion
            log.warn("Executing {} only once, since retryBackoffMs={} is larger than total timeoutMs={}",
                    descriptionStr, retryBackoffMs, timeoutMs);
```

rhauch reviewed (commented):
Thanks for the update, @philipnee. I have some very nit-picky suggestions around logging and error messages, which IMO are important to get right so that Connect users know very clearly what's happened when looking at logs or errors.

I also found one potential logic bug in the edge case when the remaining time to retry is smaller than the retry backoff time.

I think if we fix these and the build passes, we're good!

philipnee committed:
Apply suggestions from code review

Co-authored-by: Randall Hauch <rhauch@gmail.com>

philipnee committed:
retryEndOffsets should be public

- documentation

philipnee committed:
update retryEndOffsets

philipnee committed:
correcting test

philipnee committed:
validating retry and timeout

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/RetryUtil.java:
```diff
@@ -0,0 +1,109 @@
+        }
+
+        if (retryBackoffMs < 0) {
+            log.warn("Invalid retryBackoffMs, must be non-negative but got {}. 0 will be used instead",
+                    retryBackoffMs);
```
I don't think it's necessary to warn in this case. Maybe debug, but not warn. If callers are concerned, they can check the parameters before calling the method.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/RetryUtil.java:
```diff
@@ -0,0 +1,109 @@
+                .orElse(0L);
+
+        if (retryBackoffMs >= timeoutMs) {
+            log.warn("Executing {} only once, since retryBackoffMs={} is larger than total timeoutMs={}",
+                    descriptionStr, retryBackoffMs, timeoutMs);
```
I don't think it's necessary to warn in this case. Maybe debug, but not warn. If callers are concerned, they can check the parameters before calling the method.

philipnee replied:
make sense

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/RetryUtil.java:
```diff
@@ -0,0 +1,109 @@
+     * @param callable          the function to execute
+     * @param description       supplier that provides custom message for logging purpose
+     * @param timeoutDuration   timeout duration; must not be null or negative
+     * @param retryBackoffMs    the number of milliseconds to delay upon receiving a
+     *                          {@link org.apache.kafka.connect.errors.RetriableException} before retrying again;
+     *                          must be 0 or more
```
```suggestion
     * @param retryBackoffMs    the number of milliseconds to delay upon receiving a
     *                          {@link org.apache.kafka.connect.errors.RetriableException} before retrying again
```

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/RetryUtil.java:
```diff
@@ -0,0 +1,109 @@
+     *
+     * @param callable          the function to execute
+     * @param description       supplier that provides custom message for logging purpose
+     * @param timeoutDuration   timeout duration; must not be null or negative
```
```suggestion
     * @param timeoutDuration   timeout duration; must not be null
```

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/RetryUtil.java:
```diff
@@ -0,0 +1,109 @@
+     * be used for logging.  For example, "list offsets". If the supplier is null or the supplied string is
+     * null, {@code callable} will be used as the default string.
+     *
+     * <p>If {@code timeoutDuration} is set to 0, the task will be
+     * executed exactly once.  If {@code timeoutDuration} is less than {@code retryBackoffMs}, the callable will be
+     * executed only once.
```
How about:
```suggestion
     * <p>The task will be executed at least once. No retries will be performed 
     * if {@code timeoutDuration} is 0 or negative, or if {@code timeoutDuration} is less than {@code retryBackoffMs}.
```

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/RetryUtil.java:
```diff
@@ -0,0 +1,109 @@
+     * executed exactly once.  If {@code timeoutDuration} is less than {@code retryBackoffMs}, the callable will be
+     * executed only once.
+     *
+     * <p>If {@code retryBackoffMs} is set to 0, no wait will happen in between the retries.
```
```suggestion
     * <p>A {@code retryBackoffMs} that is negative or zero will result in no delays between retries.
```

rhauch reviewed (commented):
Thanks, @philipnee. Some nits per our offline discussion.

philipnee committed:
Apply suggestions from code review

Co-authored-by: Randall Hauch <rhauch@gmail.com>

philipnee committed:
make warning message debug instead

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/RetryUtil.java:
```diff
@@ -0,0 +1,107 @@
+            retryBackoffMs = 0;
+        }
+
+        if (timeoutMs <= 0 || retryBackoffMs >= timeoutMs) {
+            // no retry
+            return callable.call();
+        }
+
+        long end = System.currentTimeMillis() + timeoutMs;
+        int attempt = 0;
+        Throwable lastError = null;
+        while (System.currentTimeMillis() <= end) {
```
I think there's an edge case where `timeoutMs` is positive but small enough that the condition on line 77 is not met but the while loop on line 85 is not satisfied because the end time has already passed. In this edge case, we might not call the callable function (even once).

One option is to change the while loop to be a do-while loop so that we always go through one loop. Another option is to compute the remaining time before line 77 and not update it before the while loop. Either would work, but one of the options may require fewer duplicated lines.

philipnee replied:
I see, so if the timeoutMs is 1ms, and let's say, it takes 2ms to reach the top of the while loop.  In this case, the function will skip the loop.  

I think using a do..while loop would make sense, and we can also eliminate the duplicated callable.call() above; however, it will now be throwing ConnectException instead of the RetriableException if line 77 is satisfied.  But I think it is okay?

philipnee replied:
It's a really good catch, I think if it skips the while loop, a NPE will be thrown when it is trying to run lastError.getMessage().  In this case, lastError is null.

rhauch replied:
Ideally we'd not wrap the exception if there are no retries, so I guess it just depends on how hard it is to make that work.

philipnee replied:
I see - I should then keep the line 77 to handle special cases.

rhauch reviewed (commented):
Thanks, @philipnee. Everything looks good, but I just found a really unlikely but possible edge case that we should address. Sorry I didn't notice it earlier.

philipnee committed:
Handling small edge case when no retry might happen for small timeout

philipnee committed:
Refactor retry logic

clean up the sleep logic, add tests, and logging

philipnee committed:
handing timeoutDuration and backoffMs edge cases

rhauch reviewed (approved):
Thanks, @philipnee! LGTM.

@ijuma, do you want to double-check the reversion of [KAFKA-12339](https://issues.apache.org/jira/browse/KAFKA-12339) / #10152?
